File: tests/example-playwright-test.js
#!/usr/bin/env node

/**
 * JavaScript Playwright Test Template for JMeter Batch Framework
 * 
 * ✅ k6-Compatible API! Easy migration to k6's browser module later!
 * ✅ Writes results in JTL format compatible with merge-results Lambda
 * 
 * Usage:
 *   This script is automatically invoked by the framework when a .js test is detected.
 *   Environment variables are set by entrypoint.sh:
 *     - RESULTS_BUCKET: S3 bucket for results
 *     - CONTAINER_ID: Container identifier
 *     - RUN_ID: Test run identifier
 *     - TEST_ID: Test identifier
 *     - RESULTS_PREFIX: S3 prefix for results
 */

const { chromium } = require('playwright');
const fs = require('fs');
const { S3Client, PutObjectCommand } = require('@aws-sdk/client-s3');

// Get environment variables (set by entrypoint.sh)
const RESULTS_BUCKET = process.env.RESULTS_BUCKET || '';
const CONTAINER_ID = process.env.CONTAINER_ID || '0';
const RUN_ID = process.env.RUN_ID || 'unknown';
const TEST_ID = process.env.TEST_ID || 'unknown';
const RESULTS_PREFIX = process.env.RESULTS_PREFIX || 'results';

// Results file path (must match what merge-results expects)
const RESULTS_FILE = `/tmp/results-${CONTAINER_ID}.jtl`;

console.log('========================================');
console.log('JavaScript Playwright Test Execution (k6-ready!)');
console.log('========================================');
console.log(`Test ID: ${TEST_ID}`);
console.log(`Run ID: ${RUN_ID}`);
console.log(`Container ID: ${CONTAINER_ID}`);
console.log(`Results File: ${RESULTS_FILE}`);
console.log('========================================\n');

/**
 * Write a single result record in JMeter JTL CSV format
 * This format is compatible with merge-results Lambda and JMeter's standard output
 */
function writeJTLRecord(stream, label, elapsedMs, success, responseCode = 200, url = '', errorMsg = '') {
    const record = [
        Date.now(),                                    // timeStamp (ms since epoch)
        Math.round(elapsedMs),                        // elapsed (duration in ms)
        label,                                         // label (step/transaction name)
        responseCode,                                  // responseCode (HTTP status)
        success ? 'OK' : (errorMsg || 'FAILED'),      // responseMessage
        `Playwright-JS-${CONTAINER_ID}`,             // threadName (virtual user ID)
        'text',                                        // dataType
        success ? 'true' : 'false',                   // success (pass/fail)
        0,                                             // bytes (response size)
        url                                            // URL
    ].join(',');
    
    stream.write(record + '\n');
}

/**
 * Example Playwright test - Replace this with your actual test logic
 * 
 * This example demonstrates:
 * - Navigation
 * - Form filling
 * - Button clicking
 * - Element waiting
 * - Error handling
 * - JTL result writing
 * 
 * NOTE: This API is very similar to k6's browser module!
 * Migration to k6 later = minimal code changes!
 */
async function runExampleTest() {
    // Create JTL file with CSV header
    const stream = fs.createWriteStream(RESULTS_FILE);
    stream.write('timeStamp,elapsed,label,responseCode,responseMessage,threadName,dataType,success,bytes,URL\n');
    
    // Launch browser (headless mode for container execution)
    const browser = await chromium.launch({
        headless: true,
        args: [
            '--no-sandbox',
            '--disable-dev-shm-usage',
            '--disable-gpu',
            '--disable-software-rasterizer'
        ]
    });
    
    const page = await browser.newPage();
    
    try {
        // ===== Step 1: Navigate to Application =====
        const stepName = '01_Navigate_Homepage';
        console.log(`[STEP] ${stepName}`);
        const start1 = Date.now();
        
        try {
            const response = await page.goto('https://example.com', { 
                waitUntil: 'domcontentloaded' 
            });
            const elapsed1 = Date.now() - start1;
            
            writeJTLRecord(
                stream, stepName, elapsed1,
                response.ok(), response.status(), response.url()
            );
            console.log(`  ✅ Navigated to homepage (${Math.round(elapsed1)}ms)`);
            
        } catch (error) {
            const elapsed1 = Date.now() - start1;
            writeJTLRecord(
                stream, stepName, elapsed1,
                false, 500, 'https://example.com', error.message
            );
            console.log(`  ❌ Navigation failed: ${error.message}`);
            throw error;
        }
        
        // ===== Step 2: Click "More information" Link =====
        const stepName2 = '02_Click_More_Info';
        console.log(`[STEP] ${stepName2}`);
        const start2 = Date.now();
        
        try {
            await page.click('a:has-text("More information")', { timeout: 5000 });
            await page.waitForLoadState('networkidle');
            const elapsed2 = Date.now() - start2;
            
            writeJTLRecord(
                stream, stepName2, elapsed2,
                true, 200, page.url()
            );
            console.log(`  ✅ Clicked link (${Math.round(elapsed2)}ms)`);
            
        } catch (error) {
            const elapsed2 = Date.now() - start2;
            writeJTLRecord(
                stream, stepName2, elapsed2,
                false, 500, page.url(), error.message
            );
            console.log(`  ❌ Click failed: ${error.message}`);
            throw error;
        }
        
        // ===== Step 3: Verify Page Content =====
        const stepName3 = '03_Verify_Content';
        console.log(`[STEP] ${stepName3}`);
        const start3 = Date.now();
        
        try {
            const content = await page.content();
            const expectedText = 'Example Domain';
            
            if (content.includes(expectedText)) {
                const elapsed3 = Date.now() - start3;
                writeJTLRecord(
                    stream, stepName3, elapsed3,
                    true, 200, page.url()
                );
                console.log(`  ✅ Content verified (${Math.round(elapsed3)}ms)`);
            } else {
                const elapsed3 = Date.now() - start3;
                const errorMsg = `Expected text "${expectedText}" not found`;
                writeJTLRecord(
                    stream, stepName3, elapsed3,
                    false, 500, page.url(), errorMsg
                );
                console.log(`  ❌ Content verification failed`);
                throw new Error(errorMsg);
            }
            
        } catch (error) {
            const elapsed3 = Date.now() - start3;
            writeJTLRecord(
                stream, stepName3, elapsed3,
                false, 500, page.url(), error.message
            );
            console.log(`  ❌ Verification failed: ${error.message}`);
            throw error;
        }
        
        console.log('\n✅ Test completed successfully');
        return 0;
        
    } catch (error) {
        console.log(`\n❌ Test failed: ${error.message}`);
        return 1;
        
    } finally {
        await browser.close();
        stream.end();
    }
}

/**
 * Upload results file to S3 (same location as JMeter results)
 * Note: Upload is also handled by entrypoint.sh, this is optional
 */
async function uploadResultsToS3() {
    if (!RESULTS_BUCKET) {
        console.log('⚠️  RESULTS_BUCKET not set, skipping S3 upload');
        return;
    }
    
    if (!fs.existsSync(RESULTS_FILE)) {
        console.log(`⚠️  Results file not found: ${RESULTS_FILE}`);
        return;
    }
    
    try {
        const s3 = new S3Client({});
        const s3Key = `${RESULTS_PREFIX}/container-${CONTAINER_ID}/results-${CONTAINER_ID}.jtl`;
        
        console.log('\n[UPLOAD] Uploading results to S3...');
        console.log(`  Bucket: ${RESULTS_BUCKET}`);
        console.log(`  Key: ${s3Key}`);
        
        const fileContent = fs.readFileSync(RESULTS_FILE);
        
        await s3.send(new PutObjectCommand({
            Bucket: RESULTS_BUCKET,
            Key: s3Key,
            Body: fileContent
        }));
        
        console.log(`✅ Results uploaded: s3://${RESULTS_BUCKET}/${s3Key}`);
        
    } catch (error) {
        console.log(`❌ Failed to upload results: ${error.message}`);
    }
}

/**
 * Main execution
 */
(async () => {
    console.log('🚀 Starting JavaScript Playwright test execution\n');
    
    // Run the test
    const exitCode = await runExampleTest();
    
    // Results are written to file during test execution
    console.log(`\n📊 Results written to: ${RESULTS_FILE}`);
    
    // Show result summary
    if (fs.existsSync(RESULTS_FILE)) {
        const content = fs.readFileSync(RESULTS_FILE, 'utf8');
        const lines = content.split('\n').filter(line => line.trim());
        const total = lines.length - 1; // Subtract header
        
        console.log(`📊 Total steps executed: ${total}`);
        
        if (total > 0) {
            const successes = lines.slice(1).filter(line => line.includes(',true,')).length;
            const failures = total - successes;
            console.log(`   ✅ Passed: ${successes}`);
            console.log(`   ❌ Failed: ${failures}`);
        }
    }
    
    // Upload is handled by entrypoint.sh, but we can also do it here if needed
    // await uploadResultsToS3();
    
    console.log(`\n[EXIT] Test exiting with code: ${exitCode}`);
    process.exit(exitCode);
})();

File: tests/example-playwright-test.py
#!/usr/bin/env python3
"""
Playwright Test Template for JMeter Batch Framework
Writes results in JTL format compatible with merge-results Lambda

Usage:
  This script is automatically invoked by the framework when a .py test is detected.
  Environment variables are set by entrypoint.sh:
    - RESULTS_BUCKET: S3 bucket for results
    - CONTAINER_ID: Container identifier
    - RUN_ID: Test run identifier
    - TEST_ID: Test identifier
    - RESULTS_PREFIX: S3 prefix for results
"""

import os
import sys
import time
import csv
import boto3
from playwright.sync_api import sync_playwright

# Get environment variables (set by entrypoint.sh)
RESULTS_BUCKET = os.environ.get('RESULTS_BUCKET', '')
CONTAINER_ID = os.environ.get('CONTAINER_ID', '0')
RUN_ID = os.environ.get('RUN_ID', 'unknown')
TEST_ID = os.environ.get('TEST_ID', 'unknown')
RESULTS_PREFIX = os.environ.get('RESULTS_PREFIX', 'results')

# Results file path (must match what merge-results expects)
RESULTS_FILE = f'/tmp/results-{CONTAINER_ID}.jtl'

print(f"========================================")
print(f"Playwright Test Execution")
print(f"========================================")
print(f"Test ID: {TEST_ID}")
print(f"Run ID: {RUN_ID}")
print(f"Container ID: {CONTAINER_ID}")
print(f"Results File: {RESULTS_FILE}")
print(f"========================================\n")


def write_jtl_record(writer, label, elapsed_ms, success, response_code=200, url='', error_msg=''):
    """
    Write a single result record in JMeter JTL CSV format.
    
    This format is compatible with merge-results Lambda and JMeter's standard output.
    """
    writer.writerow({
        'timeStamp': int(time.time() * 1000),  # Milliseconds since epoch
        'elapsed': int(elapsed_ms),             # Duration in milliseconds
        'label': label,                         # Step/transaction name
        'responseCode': response_code,          # HTTP status or custom code
        'responseMessage': 'OK' if success else (error_msg or 'FAILED'),
        'threadName': f'Playwright-{CONTAINER_ID}',  # Virtual user identifier
        'dataType': 'text',                     # Data type
        'success': 'true' if success else 'false',  # Pass/fail status
        'bytes': 0,                             # Response size (optional)
        'URL': url                              # URL accessed
    })


def run_example_test():
    """
    Example Playwright test - Replace this with your actual test logic.
    
    This example demonstrates:
    - Navigation
    - Form filling
    - Button clicking
    - Element waiting
    - Error handling
    - JTL result writing
    """
    with open(RESULTS_FILE, 'w', newline='') as f:
        # JTL CSV format (same as JMeter)
        fieldnames = ['timeStamp', 'elapsed', 'label', 'responseCode',
                     'responseMessage', 'threadName', 'dataType', 'success',
                     'bytes', 'URL']
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        
        with sync_playwright() as p:
            # Launch browser (headless mode for container execution)
            browser = p.chromium.launch(
                headless=True,
                args=[
                    '--no-sandbox',
                    '--disable-dev-shm-usage',
                    '--disable-gpu',
                    '--disable-software-rasterizer'
                ]
            )
            
            page = browser.new_page()
            
            try:
                # ===== Step 1: Navigate to Application =====
                step_name = '01_Navigate_Homepage'
                print(f"[STEP] {step_name}")
                start = time.time()
                
                try:
                    response = page.goto('https://example.com', wait_until='domcontentloaded')
                    elapsed = (time.time() - start) * 1000
                    
                    write_jtl_record(
                        writer, step_name, elapsed,
                        success=response.ok,
                        response_code=response.status,
                        url=response.url
                    )
                    print(f"  ✅ Navigated to homepage ({int(elapsed)}ms)")
                    
                except Exception as e:
                    elapsed = (time.time() - start) * 1000
                    write_jtl_record(
                        writer, step_name, elapsed,
                        success=False,
                        response_code=500,
                        url='https://example.com',
                        error_msg=str(e)
                    )
                    print(f"  ❌ Navigation failed: {e}")
                    raise
                
                # ===== Step 2: Click "More information" Link =====
                step_name = '02_Click_More_Info'
                print(f"[STEP] {step_name}")
                start = time.time()
                
                try:
                    page.click('a:has-text("More information")', timeout=5000)
                    page.wait_for_load_state('networkidle')
                    elapsed = (time.time() - start) * 1000
                    
                    write_jtl_record(
                        writer, step_name, elapsed,
                        success=True,
                        response_code=200,
                        url=page.url
                    )
                    print(f"  ✅ Clicked link ({int(elapsed)}ms)")
                    
                except Exception as e:
                    elapsed = (time.time() - start) * 1000
                    write_jtl_record(
                        writer, step_name, elapsed,
                        success=False,
                        response_code=500,
                        error_msg=str(e)
                    )
                    print(f"  ❌ Click failed: {e}")
                    raise
                
                # ===== Step 3: Verify Page Content =====
                step_name = '03_Verify_Content'
                print(f"[STEP] {step_name}")
                start = time.time()
                
                try:
                    # Check if expected content is present
                    content = page.content()
                    expected_text = 'Example Domain'
                    
                    if expected_text in content:
                        elapsed = (time.time() - start) * 1000
                        write_jtl_record(
                            writer, step_name, elapsed,
                            success=True,
                            response_code=200,
                            url=page.url
                        )
                        print(f"  ✅ Content verified ({int(elapsed)}ms)")
                    else:
                        elapsed = (time.time() - start) * 1000
                        write_jtl_record(
                            writer, step_name, elapsed,
                            success=False,
                            response_code=500,
                            error_msg=f'Expected text "{expected_text}" not found'
                        )
                        print(f"  ❌ Content verification failed")
                        raise AssertionError(f'Expected text "{expected_text}" not found')
                        
                except Exception as e:
                    elapsed = (time.time() - start) * 1000
                    write_jtl_record(
                        writer, step_name, elapsed,
                        success=False,
                        response_code=500,
                        error_msg=str(e)
                    )
                    print(f"  ❌ Verification failed: {e}")
                    raise
                
                print("\n✅ Test completed successfully")
                return 0
                
            except Exception as e:
                print(f"\n❌ Test failed: {e}")
                return 1
                
            finally:
                browser.close()


def upload_results_to_s3():
    """Upload results file to S3 (same location as JMeter results)"""
    if not RESULTS_BUCKET:
        print("⚠️  RESULTS_BUCKET not set, skipping S3 upload")
        return
    
    if not os.path.exists(RESULTS_FILE):
        print(f"⚠️  Results file not found: {RESULTS_FILE}")
        return
    
    try:
        s3 = boto3.client('s3')
        s3_key = f"{RESULTS_PREFIX}/container-{CONTAINER_ID}/results-{CONTAINER_ID}.jtl"
        
        print(f"\n[UPLOAD] Uploading results to S3...")
        print(f"  Bucket: {RESULTS_BUCKET}")
        print(f"  Key: {s3_key}")
        
        s3.upload_file(RESULTS_FILE, RESULTS_BUCKET, s3_key)
        
        print(f"✅ Results uploaded: s3://{RESULTS_BUCKET}/{s3_key}")
        
    except Exception as e:
        print(f"❌ Failed to upload results: {e}")


if __name__ == '__main__':
    print("🚀 Starting Playwright test execution\n")
    
    # Run the test
    exit_code = run_example_test()
    
    # Results are written to file during test execution
    print(f"\n📊 Results written to: {RESULTS_FILE}")
    
    # Show result summary
    if os.path.exists(RESULTS_FILE):
        with open(RESULTS_FILE, 'r') as f:
            lines = f.readlines()
            total = len(lines) - 1  # Subtract header
            print(f"📊 Total steps executed: {total}")
            
            # Count successes and failures
            if total > 0:
                successes = sum(1 for line in lines[1:] if ',true,' in line)
                failures = total - successes
                print(f"   ✅ Passed: {successes}")
                print(f"   ❌ Failed: {failures}")
    
    # Upload is handled by entrypoint.sh, but we can also do it here if needed
    # upload_results_to_s3()
    
    print(f"\n[EXIT] Test exiting with code: {exit_code}")
    sys.exit(exit_code)